base/testfiles-lthooks/lthooks-033.lvt

```diff
@@ -0,0 +1,140 @@
+    % label
+    \AddToHook{\hookname}[label]{\typeout{label-hashes(#1,#2)}}
+    \AddToHookWithArguments{\hookname}[label]{\typeout{label-args(#1,#2)}}
+    \AddToHook{\hookname}[label]{\typeout{label-hashes(#1,#2)}}
+    \AddToHookWithArguments{\hookname}[label]{\typeout{label-args(#1,#2)}}
+    
+    \UseHook{\hookname}{foo}{bar}
+  }
+
+%
+
+\TEST{Zero arguments}
+  {
+    \def\hookname{test-0}
+    \NewHookWithArguments{\hookname}{0}
+    % top-level
+    \AddToHook{\hookname}{\typeout{top-hashes(#1,#2)}}
+    \AddToHookWithArguments{\hookname}{\typeout{top-args(#1,#2)}}
+    \AddToHook{\hookname}{\typeout{top-hashes(#1,#2)}}
+    \AddToHookWithArguments{\hookname}{\typeout{top-args(#1,#2)}}
+    % next
+    \AddToHookNext{\hookname}{\typeout{next-hashes(#1,#2)}}
+    \AddToHookNextWithArguments{\hookname}{\typeout{next-args(#1,#2)}}
+    \AddToHookNext{\hookname}{\typeout{next-hashes(#1,#2)}}
+    \AddToHookNextWithArguments{\hookname}{\typeout{next-args(#1,#2)}}
+    % label
+    \AddToHook{\hookname}[label]{\typeout{label-hashes(#1,#2)}}
+    \AddToHookWithArguments{\hookname}[label]{\typeout{label-args(#1,#2)}}
+    \AddToHook{\hookname}[label]{\typeout{label-hashes(#1,#2)}}
+    \AddToHookWithArguments{\hookname}[label]{\typeout{label-args(#1,#2)}}
+    
+    \UseHook{\hookname}{foo}{bar}
+  }
+
+%
+
+\TEST{Ten arguments}
+  {
+    % Too many arguments, then LaTeX defines with zero, then
+    % \AddToHookWithArguments complains...  Maybe should define with 9?
+    \NewHookWithArguments{test-10}{10}
+    \AddToHookWithArguments{test-10}{\typeout{10-top-level(#1,#2)}}
+    \AddToHookWithArguments{test-10}[label]{\typeout{10-label(#1,#2)}}
+    \AddToHookNextWithArguments{test-10}{\typeout{10-next(#1,#2)}}
+    \UseHook{test-10}{foo}{bar}%
+      {3}{4}{5}{6}{7}{8}{9}
+  }
+
+%
+
+\TEST{No arguments}
+  {
+    \NewHook{no-args}
+    \AddToHook{no-args}{\typeout{no-top-level(#1,#2)}} % ok, adds ##1, ##2
+    \AddToHook{no-args}[label]{\typeout{no-label(#1,#2)}} % ok, adds ##1, ##2
+    \AddToHookNext{no-args}{\typeout{no-next(#1,#2)}} % ok, adds ##1, ##2
+    
+    \AddToHookWithArguments{no-args}{\typeout{no-top-level(#1,#2)}} % error, adds ##1, ##2
+    \AddToHookWithArguments{no-args}[label]{\typeout{no-label(#1,#2)}} % error, adds ##1, ##2
+    \AddToHookNextWithArguments{no-args}{\typeout{no-next(#1,#2)}} % error, adds ##1, ##2
+    \UseHook{no-args}
+  }
+
+%
+
+\TEST{One-time hook with arguments}
+  {
+    \NewHookWithArguments{use-once}{2}
+    \AddToHookWithArguments{use-once}{\typeout{once(#1,#2)}}
+    \UseOneTimeHookWithArguments{use-once}{foo}{bar}
+    \AddToHookWithArguments{use-once}{\typeout{twice(#1,#2)}}
+  }
+
+%
+
+\TEST{Add with arguments before hook is declared}
+  {
+    \AddToHookWithArguments{not-declared}{\typeout{args(#1,#2)}}
+    \NewHookWithArguments{not-declared}{2}
+    \AddToHookWithArguments{not-declared}{\typeout{more-args(#1,#2)}}
+    \UseHookWithArguments{not-declared}{foo}{bar}
+  }
```

Comment:
what happens if there is a mix and match? ie if there is also
```
\AddToHook{not-declared}{\typeout{not args: #1 #2}}
```
before the hook is declared

Reply:
It works (now :).  Somehow this went through the tests I had before.  Hashes added with `\AddToHook` remain as parameter tokens, and hashes added with `\AddToHookWithArguments` are (unless doubled) treated as arguments.  That's the idea, at least.